pulsar-common/src/main/java/org/apache/pulsar/common/util/RateLimiter.java

```diff
@@ -257,14 +271,15 @@ public synchronized TimeUnit getRateTimeUnit() {
     }
 
     synchronized void renew() {
-        acquiredPermits = isDispatchRateLimiter ? Math.max(0, acquiredPermits - permits) : 0;
+        acquiredPermits = isDispatchOrPrecisePublishRateLimiter ? Math.max(0, acquiredPermits - permits) : 0;
         if (permitUpdater != null) {
             long newPermitRate = permitUpdater.get();
             if (newPermitRate > 0) {
                 setRate(newPermitRate);
             }
         }
-        if (rateLimitFunction != null) {
+        // release the back-pressure by applying the rateLimitFunction only when there are available permits
+        if (rateLimitFunction != null && this.getAvailablePermits() > 0) {
```

Comment:
I'm not sure if this is necessary
Changed existing behavior

Reply:
It is necessary, without it every renew call the function will release the throttle.

Not having it defeats the whole purpose of throttling a connection.

Reply:
is that still relevant?
and can you also review the test refactor?